### PR TITLE
Attempt to reconnect to streams that are terminated with unexpected errors

### PR DESCRIPTION
**Pull request by Sovietaced**
#### What type of PR is this?
This is a bug fix

#### What this PR does / why we need it:
If the [job service subscription manager goroutine](https://github.com/armadaproject/armada/blob/master/internal/jobservice/eventstojobs/manage_subs.go#L271-L338) receives an error (besides an EOF) from the Armada API it will be stuck in a terminal state reading the same error from the stream and writing it to the database, endlessly.  That is because a stream error is terminal. The server ends the stream with an error and any calls to read from the stream will return the same error over and over again. This is a common mistake with gRPC stream client logic unfortunately.

We specifically saw this become an issue when the Armada API server was propagating redis connection pool errors back over the stream to the jobservice subscriptions under heavy concurrent load. 

This pull request updates the behavior to treat the stream receive errors as terminal and introduces a reconnection mechanism to overcome transient armada API issues that is fairly simple for now. It reconnects with a linear backoff. 

Stream creation errors will still result in halting subscription goroutines. 

#### Which issue(s) this PR fixes:
Fixes https://github.com/armadaproject/armada/issues/3553

#### Special notes for your reviewer:

We considered updating the goroutine logic to return when an error is received on the stream but we were worried that this would remove any materialized state about the jobs from the Postgres database. More specifically we were concerned that if the subscription was restarted, clients of the job service API might see responses that jump backwards in time, back towards initial message IDs. As such, we felt it was best to keep the goroutine and the state in Postgres alive.

### CODE REVIEW

File: internal/jobservice/eventstojobs/manage_subs_test.go
Comment (Sovietaced) [inline internal/jobservice/eventstojobs/manage_subs_test.go:45]: Updates the mock server to behave like a real server would.

File: internal/jobservice/eventstojobs/manage_subs_test.go
Comment (Sovietaced) [inline internal/jobservice/eventstojobs/manage_subs_test.go:78]: Refactored so that we can mock reconnecting and acquiring new streams.

**Review by richscott — DISMISSED** [review]
Looks good - great fix!

**Review by ClifHouck — APPROVED** [review]
Seems OK to me as well. Should be fine to merge as long as tests pass.

**Comment by Sovietaced** [comment]
Lint issue should be fixed now

**Comment by Sovietaced** [comment]
@ClifHouck Can you please re-approve so the CI runs

**Review by ClifHouck — APPROVED** [review]

**Comment by Sovietaced** [comment]
ping @richscott or any other maintainer. Just hit this issue today in production, however this time caused by what seems like some network error. 

```
could not obtain job set event message, retrying  \u001b[31merror\u001b[0m=\"rpc error: code = Unavailable desc = error reading from server: EOF
```

**Review by richscott — APPROVED** [review]

**Pull request merged**
